core/drivers/crypto/hisilicon/hpre_rsa.c

```diff
@@ -0,0 +1,1076 @@
+}
+
+static TEE_Result rsa_pkcs_decrypt(struct drvcrypt_rsa_ed *rsa_data)
+{
+	uint32_t n_bytes = rsa_data->key.n_size;
+	struct drvcrypt_rsa_ed rsa_dec_info = *rsa_data;
+	TEE_Result ret = TEE_SUCCESS;
+
+	/* Alloc pkcs_v1.5 encode message data buf */
+	rsa_dec_info.message.data = malloc(n_bytes);
+	if (!rsa_dec_info.message.data) {
+		EMSG("Fail to alloc message data buf");
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	rsa_dec_info.message.length = n_bytes;
+	ret = rsa_nopad_decrypt(&rsa_dec_info);
+	if (ret)
+		goto free_data;
+
+	ret = rsaes_pkcs_v1_5_decode(&rsa_dec_info, rsa_data->message.data,
+				     &rsa_data->message.length);
+	if (ret)
+		EMSG("Fail to get pkcs_v1.5 decode message data");
+
+free_data:
+	free(rsa_dec_info.message.data);
+
+	return ret;
+}
+
+static TEE_Result rsa_oaep_get_seed(struct drvcrypt_rsa_ed *rsa_data,
+				    uint8_t *mask_db, uint8_t *seed)
+{
+	size_t db_len = rsa_data->key.n_size - rsa_data->digest_size - 1;
+	uint8_t mask_db_mgf[OAEP_MAX_HASH_LEN] = { };
+	size_t lhash_len = rsa_data->digest_size;
+	uint8_t *mask_seed = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	mask_seed = rsa_data->message.data + 1;
+
+	ret = mgf_process(lhash_len, mask_db, db_len, mask_db_mgf, lhash_len,
+			  rsa_data);
+	if (ret) {
+		EMSG("Fail to get mask_db mgf result");
+		return ret;
+	}
+
+	return xor_process(mask_seed, mask_db_mgf, seed, lhash_len);
+}
+
+static TEE_Result rsa_oaep_get_db(struct drvcrypt_rsa_ed *rsa_data,
+				  uint8_t *mask_db, uint8_t *seed, uint8_t *db)
+{
+	size_t db_len = rsa_data->key.n_size - rsa_data->digest_size - 1;
+	size_t lhash_len = rsa_data->digest_size;
+	uint8_t seed_mgf[OAEP_MAX_DB_LEN] = { };
+	TEE_Result ret = TEE_SUCCESS;
+
+	ret = mgf_process(lhash_len, seed, lhash_len, seed_mgf, db_len,
+			  rsa_data);
+	if (ret) {
+		EMSG("Fail to get seed mgf result");
+		return ret;
+	}
+
+	return xor_process(mask_db, seed_mgf, db, db_len);
+}
+
+static TEE_Result rsa_oaep_get_msg(struct drvcrypt_rsa_ed *rsa_data,
+				   uint8_t *db, uint8_t *out, size_t *out_len)
+{
+	size_t db_len = rsa_data->key.n_size - rsa_data->digest_size - 1;
+	size_t lhash_len = rsa_data->digest_size;
+	uint8_t hash[OAEP_MAX_HASH_LEN] = { };
+	size_t msg_len = 0;
+	size_t lp_len = 0;
+	TEE_Result ret = TEE_SUCCESS;
+
+	/* oaep db format lhash || ps zero || 01 || M */
+	ret = tee_hash_createdigest(rsa_data->hash_algo, rsa_data->label.data,
+				    rsa_data->label.length, hash, lhash_len);
+	if (ret) {
+		EMSG("Fail to get label hash");
+		return ret;
+	}
+
+	if (memcmp(hash, db, lhash_len)) {
+		EMSG("Hash is not equal");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	for (lp_len = lhash_len; lp_len < db_len; lp_len++) {
+		if (db[lp_len] != 0)
+			break;
+	}
+
+	if (lp_len == db_len) {
+		EMSG("Fail to find fixed 01 in db");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	msg_len = db_len - lp_len - 1;
+	if (msg_len > rsa_data->message.length) {
+		EMSG("Message space is not enough");
+		return TEE_ERROR_SHORT_BUFFER;
```

Comment:
Set `*out_len = msg_len` to help caller to get to required size.

Reply:
solved

Reply:
I would replace `EMSG()` with `DMSG()` at line 918 since this may not be a error case but an output buffer  length query from the caller.